### PR TITLE
Fix17713 - Reverting PR - 17649 - Make the interaction between #line and #nowarn directives consistent

### PR DESCRIPTION
**Pull request by KevinRansom**
This language feature caused bootstrapping issues for this repo introducing additional unintended warnings when building the F# bootstrapper with F# 9.0.

This was first noticed by the sourcebuildteam who have a strict bootstrapping requirement for delivery to Linux distributions.

Additionally the langversion:8 implementation produced different error output from the 8.0.* version of the compiler.

The way that we use #line and # n filename in the product is quite extensive and fragile.  It is clear that we need a more robust solution highlighted by the issue the original PR was intended to solve, this has highlighted that we do need to address the consistency issue.  Everyone on the team agreed that the current implementation that we are going back to is not ideal, but this reversion allows us to identify a better design for a fix.

The warnings by bootstrapping with F# 9.0 produced are:

 ````
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(35,39): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_decl has been constrained to be type 'ParserSpec -> ParserSpec'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(38,107): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_idents has been constrained to be type 'Identifier list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(38,107): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_idents has been constrained to be type 'Identifier list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(46,42): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_rule has been constrained to be type 'Identifier * Rule list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(46,42): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_rule has been constrained to be type 'Identifier * Rule list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(47,69): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clauses has been constrained to be type 'Rule list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(47,69): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clauses has been constrained to be type 'Rule list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(411,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optbar has been constrained to be type 'unit'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(411,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optbar has been constrained to be type 'unit'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(431,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optsemi has been constrained to be type 'unit'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(431,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optsemi has been constrained to be type 'unit'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(50,51): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clause has been constrained to be type 'Rule'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(50,51): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clause has been constrained to be type 'Rule'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(51,55): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_syms has been constrained to be type 'Identifier list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(51,55): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_syms has been constrained to be type 'Identifier list'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(51,58): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optprec has been constrained to be type 'Identifier option'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(51,58): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optprec has been constrained to be type 'Identifier option'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
  /vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(223,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype__startspec has been constrained to be type 'obj'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
##[error]/vmr/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(223,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype__startspec has been constrained to be type 'obj'. [/vmr/src/fsharp/buildtools/fsyacc/fsyacc.fsproj]
````

### CODE REVIEW

File: docs/release-notes/.Language/preview.md [hunk 1]
```diff
@@ -4,4 +4,10 @@
 
 ### Fixed
 
+* Allow extension methods without type attribute work for types from imported assemblies. ([PR #16368](https://github.com/dotnet/fsharp/pull/16368))
+* Enforce AttributeTargets on let values and functions. ([PR #16692](https://github.com/dotnet/fsharp/pull/16692))
```
Comment (vzarytovskii) [inline docs/release-notes/.Language/preview.md:8]: These shouldn't be here, but in 9 changelog

**Comment by vzarytovskii** [comment]
FYI @Martin521, it was blocking our sourcebuild pipeline (Kevin tried to fix it yesterday, but no luck apparently).

File: docs/release-notes/.Language/preview.md (hunk 1 above)
Comment (KevinRansom) [reply docs/release-notes/.Language/preview.md:8]: cool.

Comment (KevinRansom) [reply docs/release-notes/.Language/preview.md]: done

**Comment by Martin521** [comment]
> FYI @Martin521, it was blocking our sourcebuild pipeline (Kevin tried to fix it yesterday, but no luck apparently).

How can I reproduce this?
I am not sure what "bootstrapping with F# 9.0" exactly means.
It might make sense to integrate at least the changes to fsyaccdriver.fs and the *.fsi file to avoid the same problem next year.

**Comment by KevinRansom** [comment]
@Martin521 Build the bootstrapper with an F# 9.0 compiler with these changes.

**Comment by Martin521** [comment]
Ok, if that is a requirement, then this fix can never be done in one step.
It is important then, to already now insert the extra #nowarn directives into the grammar files.
This doesn't hurt, does not change the behavior or output, but enables us next year to create a fix.

**Review by Martin521 — CHANGES_REQUESTED** [review]
Suggest to not revert the additional #nowarn directives so that the original issue can be solved in F# 10.

**Comment by Martin521** [comment]
And, even more important, `#nowarn "64"` should be provisionally inserted in `fsyaccpars.fs` after line 15.
I will create a PR in the FsLexYacc repository to change the upstream source `fsyaccpars.fsy` to the same effect.

**Comment by Martin521** [comment]
Actually, with just this additional `#nowarn` in `fsyaccpars`, the original fix (#17649) will work also in F# 9 bootstrapping. So, up to the team if the fix should happen now or next year.

**Review by T-Gro — APPROVED** [review]

**Comment by KevinRansom** [comment]
@Martin521 we are going with the reversion for now, we need to have a fuller discussion on how we want nowarn to work if we decide to change it.

Here is my discussion of why I believe reversion is the way to go right now:  https://github.com/dotnet/fsharp/pull/17748#issuecomment-2353532052

**Comment by Martin521** [comment]
The ["scoped nowarn" RFC discussion thread](https://github.com/fsharp/fslang-design/discussions/786) might be an appropriate place for a fuller discussion.

**Comment by vzarytovskii** [comment]
> The ["scoped nowarn" RFC discussion thread](https://github.com/fsharp/fslang-design/discussions/786) might be an appropriate place for a fuller discussion.

Yeah, we should continue there. @KevinRansom @T-Gro we should discuss it on some of the upcoming triage sessions

**Comment by ellahathaway** [comment]
Please make sure that this flows into .NET 9.0. I believe fsharp's `refs/heads/release/dev17.12` branch flows to .NET 9.0 release branch. Thanks in advance :)

**Comment by ellahathaway** [comment]
> Please make sure that this flows into .NET 9.0. I believe fsharp's refs/heads/release/dev17.12 branch flows to .NET 9.0 release branch. Thanks in advance :)

@dotnet/fsharp - This change needs to be ported into .NET 9. What is the best way to get this backported to the `fsharp's refs/heads/release/dev17.12` branch?

**Pull request merged**